### PR TITLE
implement upload, list, and delete of the files api

### PR DESCRIPTION
**Pull request by kcberg**
https://platform.openai.com/docs/api-reference/files

The tests do the job but are not great given timing issues with file uploads.

FWIW I plan to finish files and then implement the fine-tunes api.

### CODE REVIEW

File: examples/chat_cli.rs [hunk 1]
```diff
@@ -21,7 +23,7 @@ async fn main() {
     }];
 
     loop {
-        print!("User: ");
+        print!("User[{}]: ", messages.len());
```
Comment (rellfy) [inline examples/chat_cli.rs:26]: What's the benefit that this change brings?

File: src/files.rs [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    data: Vec<File>,
+    object: String,
+}
+
+#[derive(Serialize, Builder, Debug, Clone)]
+#[builder(pattern = "owned")]
+#[builder(name = "FileUploadBuilder")]
+#[builder(setter(strip_option, into))]
+pub struct FileUploadRequest {
+    filename: String,
```
Comment (rellfy) [inline src/files.rs:36]: Could this be renamed to `file_name` instead, for consistency? Same for the `filepath` instances.

File: src/files.rs [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+            }));
+        }
+        let simple_name = upload_filepath
+            .file_name()
+            .unwrap()
+            .to_str()
+            .unwrap()
+            .to_string()
+            .clone();
+        let async_file = tokio::fs::File::open(upload_filepath).await.unwrap();
```
Comment (rellfy) [inline src/files.rs:60]: I would prefer if we got rid of the `unwrap`s here and returned errors instead.
I also plan on flattening the `ApiResponseOrError` type, as nested errors are a bit confusing IMO. But I am happy to merge this in as-is and address the `unwrap`s before the next release.

**Review by rellfy — COMMENTED** [review]
Thanks a lot for your contribution. LGTM apart from a few minor suggestions if you don't mind addressing them.

File: src/files.rs (hunk 3 above)
Comment (kcberg) [reply src/files.rs]: Agreed, the nested errors are proving to be difficult to work with. 
I tried to make it a little better but some tests are showing it's not quite working correctly.

Comment (kcberg) [reply src/files.rs]: Here is a branch with a flattened `ApiResponseOrError`. 

https://github.com/rellfy/openai/compare/master...kcberg:feature/error-handling

If you think this is the right approach for error handling I'll open a PR.

Comment (rellfy) [reply src/files.rs]: I'm happy to go with that. I was thinking to make it enum instead, but that's fine too.

**Pull request merged**